### PR TITLE
mark __cxa_allocate_exception, __cxa_free_exception and __cxa_init_primary_exception noexcept

### PR DESCRIPTION
**Pull request by itrofimow**
This patch marks `__cxa_allocate_exception`, `__cxa_free_exception` and `__cxa_init_primary_exception` `noexcept` to make these declarations compatible with libsupc++ and libc++abi. 

Aims to resolve https://github.com/libcxxrt/libcxxrt/issues/25

### CODE REVIEW

**Comment by dankm** [comment]
Looks correct to me, and on my local system with this change in place everything's working as expected and LLVM is also happy.

It's happier with #26 as well added too.

**Comment by dankm** [comment]
My only issue is (and this is me being a stickler for git conventions) the formatting of the commit message.

**Review by davidchisnall — APPROVED** [review]

**Comment by DimitryAndric** [comment]
The "Test in FreeBSD (Debug)" job failed, but not particularly due to these changes; it seems to have gone into a bootloop:

```text
2024-01-26T02:58:41.3482780Z <<BOOT>> -
2024-01-26T02:58:41.3494710Z Copyright (c) 1992-2821 The FreeBSD Project.
2024-01-26T02:58:41.3504810Z Copyright (c) 1979, 1988, 1983, 1986, 1988, 1989, 1991, 1992, 1993, 1994
2024-01-26T02:58:41.3514090Z The Regents of the University of California. All rights reserved.
2024-01-26T02:58:41.3539890Z FreeBSD is a registered trademark of The FreeBSD Foundation.
2024-01-26T02:58:41.3545150Z FreeBSD 13.2-RELEASE releng/13.2-n254617-525ecfdad597 GENERIC amd64
2024-01-26T02:58:41.3549100Z FreeBSD clang version 14.8.5 (https://github.com/Ilvm/1lvm-project .git 1lvmorg-1
2024-01-26T02:58:41.3569670Z 4.8.5-8-gc12386ae247c)
2024-01-26T02:58:41.3571640Z VT (vga): text 88x25
2024-01-26T02:58:41.3573500Z CPU: Intel(R) Core(TM) i7-8708B CPU @ 3.2@GHz (3193.84-MHz KB-class CPU)
2024-01-26T02:58:41.3575790Z Origin="GenuineIntel  Id=@x3@6aa_ Family=@x6 Model=@x3a Stepping=18
2024-01-26T02:58:41.3578100Z Features=@x1783f bff <FPU, VME ,DE,PSE,TSC.MSR,PAE MCE ,CX8,APIC,SEP.MTRR,PGE .MCA.C
2024-01-26T02:58:41.3580850Z HOV. PAT ,PSE36 , MMX,FXSR SSE ,SSE2,HTT>
2024-01-26T02:58:41.3582870Z Features2=@x5eda2283<SSE3 , PCLMULQDO, SSSE3,CX16,PCID,SSE4.1,5SE4.2,MOVBE ,POPCNT
2024-01-26T02:58:41.3584880Z /AESNI, XSAVE , OSXSAVE , AVX, RDRAND>
2024-01-26T02:58:41.3586500Z AND Features=@x20188880<SYSCALL .NX.LM>
2024-01-26T02:58:41.3588130Z AND Features2=8x121<LAHF ,ABM,Prefetch>
2024-01-26T02:58:41.3590150Z Structured Extended Features=@x842421<FSGSBASE ,AVX2, INVPCID.NFPUSG , RDSEED, CLFL
2024-01-26T02:58:41.3592120Z USHOPT>
2024-01-26T02:58:41.3593470Z real memory = 1873676288 (1823 MB)
2024-01-26T02:58:41.3595020Z avail memory = 1883921488 (957 MB)
2024-01-26T02:58:41.3597380Z Event timer  LAPIC  quality 108
2024-01-26T02:58:41.3598800Z ACPI APIC Table: <VBOX  VBOXAPIC>
2024-01-26T02:58:43.8184450Z <<BOOT>> -
2024-01-26T02:58:43.8195010Z Copyright (c) 1992-2821 The FreeBSD Project.
2024-01-26T02:58:43.8205030Z Copyright (c) 1979, 1988, 1983, 1986, 1988, 1989, 1991, 1992, 1993, 1994
2024-01-26T02:58:43.8228590Z The Regents of the University of California. All rights reserved.
2024-01-26T02:58:43.8236290Z FreeBSD is a registered trademark of The FreeBSD Foundation.
2024-01-26T02:58:43.8251120Z FreeBSD 13.2-RELEASE releng/13.2-n254617-525ecfdad597 GENERIC amd64
2024-01-26T02:58:43.8254350Z FreeBSD clang version 14.8.5 (https://github.com/Ilvm/1lvm-project .git 1lvmorg-1
2024-01-26T02:58:43.8257320Z 4.8.5-8-gc12386ae247c)
2024-01-26T02:58:43.8259750Z VT (vga): text 88x25
2024-01-26T02:58:43.8262530Z CPU: Intel(R) Core(TM) i7-8708B CPU @ 3.2@GHz (3193.84-MHz KB-class CPU)
2024-01-26T02:58:43.8265640Z Origin="GenuineIntel  Id=@x3@6aa_ Family=@x6 Model=@x3a Stepping=18
2024-01-26T02:58:43.8268750Z Features=@x1783f bff <FPU, VME ,DE,PSE,TSC.MSR,PAE MCE ,CX8,APIC,SEP.MTRR,PGE .MCA.C
2024-01-26T02:58:43.8272630Z HOV. PAT ,PSE36 , MMX,FXSR SSE ,SSE2,HTT>
2024-01-26T02:58:43.8275730Z Features2=@x5eda2283<SSE3 , PCLMULQDO, SSSE3,CX16,PCID,SSE4.1,5SE4.2,MOVBE ,POPCNT
2024-01-26T02:58:43.8279000Z /AESNI, XSAVE , OSXSAVE , AVX, RDRAND>
2024-01-26T02:58:43.8282340Z AND Features=@x20188880<SYSCALL .NX.LM>
2024-01-26T02:58:43.8286260Z AND Features2=8x121<LAHF ,ABM,Prefetch>
2024-01-26T02:58:43.8290150Z Structured Extended Features=@x842421<FSGSBASE ,AVX2, INVPCID.NFPUSG , RDSEED, CLFL
2024-01-26T02:58:43.8293530Z USHOPT>
2024-01-26T02:58:43.8296090Z real memory = 1873676288 (1823 MB)
2024-01-26T02:58:43.8298790Z avail memory = 1883921488 (957 MB)
2024-01-26T02:58:43.8301660Z Event timer  LAPIC  quality 108
2024-01-26T02:58:43.8304420Z ACPI APIC Table: <VBOX  VBOXAPIC>
2024-01-26T02:58:46.3918930Z <<BOOT>> -
...  on and on for a few hours, then ...
2024-01-26T08:55:00.6570520Z ##[error]The operation was canceled.
```

In any case I think this should indeed be the correct solution, at least for now. (In the future we might want to add a `_LIBCXXRT_NOEXCEPT` define to `cxxabi.h`, to at least use the same spelling everywhere.)

Can we also make the commit message slightly less terse?  I would suggest:

```text
Mark __cxa_allocate_exception, __cxa_free_exception and __cxa_init_primary_exception noexcept

This makes the declarations compatible with libstdc++ and libc++.
```

I don't know of a good way to shorten the initial line to <= 72 characters though, without obfuscating the meaning. :)

**Comment by itrofimow** [comment]
I've amended the commit message, could you please have a look @DimitryAndric ?

**Comment by dankm** [comment]
Can't speak for @DimitryAndric but I like this commit message better.

**Review by DimitryAndric — APPROVED** [review]
LGTM

**Comment by DimitryAndric** [comment]
Btw I can't merge this via the GitHub UI, I think @davidchisnall is the only person with permission to do that? :)

**Comment by davidchisnall** [comment]
@DimitryAndric, should be possible now.

**Comment by DimitryAndric** [comment]
Oh I think I was mistaken about the cause: it says "Required statuses must pass before merging", at least for auto-merge. Let's see if the CI jobs complete successfully first.

**Comment by davidchisnall** [comment]
If not, we should probably move FreeBSD tests back to Cirrus CI.

**Comment by DimitryAndric** [comment]
So the Ubuntu bots are done in ~10s, the Mac bots in 1m, but the FreeBSD bots are still booting? Looks like they're rather broken.

**Comment by DimitryAndric** [comment]
@davidchisnall I still can't merge this from the GitHub UI. Btw do we prefer merge commits or rebases? It should not make a difference in this case, but in general?

**Comment by davidchisnall** [comment]
I generally prefer rebases for individual commits, since a merge is just clutter.

**Comment by DimitryAndric** [comment]
I can't push this, as the repository refuses direct pushes to master. I also can't merge the pull request, since that UI is disabled for me. @davidchisnall please push the button :)

**Comment by DimitryAndric** [comment]
Ok, we need one more change. I built it locally here in the same fashion the CI builders do (e.g. using `cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=ON -DCOMPARE_TEST_OUTPUT_TO_SYSTEM_OUTPUT=OFF -G Ninja` to configure it), and there is one instance of `__cxa_init_primary_exception` that has no `throw()` specifier in `test/test_exception.cc`:

```text
FAILED: test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o
/usr/bin/c++  -I/share/dim/src/libcxxrt/src -O3 -DNDEBUG -std=gnu++11 -MD -MT test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o -MF test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o.d -o test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o -c /share/dim/src/libcxxrt/test/test_exception.cc
/share/dim/src/libcxxrt/test/test_exception.cc:337:19: error: functions that differ only in their return type cannot be overloaded
  337 |         __cxa_exception *__cxa_init_primary_exception(
      |         ~~~~~~~~~~~~~~~~~^
/usr/include/c++/v1/__exception/exception_ptr.h:36:47: note: previous declaration is here
   36 | _LIBCPP_OVERRIDABLE_FUNC_VIS __cxa_exception* __cxa_init_primary_exception(
      |                              ~~~~~~~~~~~~~~~~ ^
1 error generated.
```

**Comment by DimitryAndric** [comment]
Ah, and there is another problem in `test_exception.cc`: the hand-rolled `__cxa_*` declarations aren't wrapped in a `__cxxabiv1` namespace, which in combination with libc++ 18's `<exception>` results in:

```text
FAILED: test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o
/usr/bin/c++  -I/share/dim/src/libcxxrt/src -O3 -DNDEBUG -std=gnu++11 -MD -MT test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o -MF test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o.d -o test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o -c /share/dim/src/libcxxrt/test/test_exception.cc
/share/dim/src/libcxxrt/test/test_exception.cc:337:19: error: functions that differ only in their return type cannot be overloaded
  337 |         __cxa_exception *__cxa_init_primary_exception(
      |         ~~~~~~~~~~~~~~~~~^
/usr/include/c++/v1/__exception/exception_ptr.h:36:47: note: previous declaration is here
   36 | _LIBCPP_OVERRIDABLE_FUNC_VIS __cxa_exception* __cxa_init_primary_exception(
      |                              ~~~~~~~~~~~~~~~~ ^
1 error generated.
```

I've testedthe following diff to fix all of that:

```diff
diff --git a/test/test_exception.cc b/test/test_exception.cc
index f27f761..06cfd11 100644
--- a/test/test_exception.cc
+++ b/test/test_exception.cc
@@ -324,6 +324,7 @@ void test_uncaught_exceptions()
 	catch (...) {}
 }
 
+namespace __cxxabiv1 {
 extern "C"
 {
 	struct __cxa_exception;
@@ -335,11 +336,14 @@ extern "C"
 	void __cxa_decrement_exception_refcount(void* thrown_exception);
 
 	__cxa_exception *__cxa_init_primary_exception(
-		void *object, std::type_info* tinfo, void (*dest)(void *));
+		void *object, std::type_info* tinfo, void (*dest)(void *)) throw();
 
 	void __cxa_rethrow_primary_exception(void* thrown_exception);
 	void *__cxa_current_primary_exception();
 }
+}
+
+using namespace __cxxabiv1;
 
 class ExceptionPtr final {
 public:
```

**Comment by itrofimow** [comment]
Thank you for doing that @DimitryAndric.
I can add your patch to this PR during the weekend if that is not urgent, and if it is i can do it later today

**Comment by itrofimow** [comment]
No matter what we do, we clash with either libc++ definition or libstdc++ definition, but we only clash if we include `exception`.

I moved the test into separate file, which doesn't have that include, and tested with
`gcc-12` + `libstdc++12`, `clang-16` + `libstdc++12`, `clang-16` + `libc++17`, `clang-16` + `libc++18`, and all these configurations work (FWIW clang + libstdc++ is angry about `terminate` not being `noexcept`, as it clashes with https://github.com/gcc-mirror/gcc/blob/releases/gcc-12.3.0/libstdc%2B%2B-v3/include/bits/c%2B%2Bconfig#L311) for me locally even with `throw()` sed-ed into `noexcept`.

@DimitryAndric could you please validate that your setup works with the last 2 commits from this PR included?

**Comment by itrofimow** [comment]
Hi!
Is this only blocked on CI misbehaving for FreeBSD, or is there anything else that should be done here?

@davidchisnall @DimitryAndric

**Comment by davidchisnall** [comment]
I'm still blocked on @emaste, since he's the only person with the required permissions to fix the CI.

**Comment by emaste** [comment]
Sorry, I'll look at this later today

**Comment by davidchisnall** [comment]
If you can grant me owner permission on the org, I can fix it.

**Comment by davidchisnall** [comment]
CI is now green, so please rebase this and fix the conflicts.

**Pull request merged**
